File: src/elixir/svd/test/exporters/json_exporter_test.exs
defmodule SVD.Exporters.JSONExporterTest do
  use ExUnit.Case, async: true

  alias SVD.Exporters.Formats.JSONExporter
  alias SVD.Models.StandardisedVesselDataset
  alias SVD.TestSupport.Faker
  alias SVD.Validators.{StandardisedVesselDatasetValidator, ValidatorException}

  defmodule UnknownStruct do
    defstruct [:value]
  end

  setup do
    %{exporter: JSONExporter.new(StandardisedVesselDatasetValidator)}
  end

  test "default struct can be created" do
    exporter = struct(JSONExporter)
    assert exporter.validator == StandardisedVesselDatasetValidator
  end

  test "export nil svd", %{exporter: exporter} do
    assert {:error, "SVD cannot be nil"} = JSONExporter.export_async(exporter, nil)
  end

  test "export invalid svd", %{exporter: exporter} do
    assert {:error, %ValidatorException{} = exception} =
             JSONExporter.export_async(exporter, Faker.invalid_svd())

    assert ValidatorException.has_errors?(exception)
  end

  test "export valid svd", %{exporter: exporter} do
    assert {:ok, content} = JSONExporter.export_async(exporter, Faker.valid_svd())

    assert byte_size(content.data) > 0
    assert String.contains?(content.file_name, Faker.valid_svd().general.imo)
    assert content.content_type == "application/json"
  end

  test "export minimal valid data", %{exporter: exporter} do
    minimal = %StandardisedVesselDataset{general: Faker.valid_general_information()}
    assert {:ok, content} = JSONExporter.export_async(exporter, minimal)
    assert byte_size(content.data) > 0
  end

  test "valid json format", %{exporter: exporter} do
    assert {:ok, content} = JSONExporter.export_async(exporter, Faker.valid_svd())
    assert {:ok, payload} = Jason.decode(content.data)

    assert Map.has_key?(payload, "general")
    assert Map.has_key?(payload, "portAndRoute")
    assert Map.has_key?(payload, "arrivalTimes")
    assert Map.has_key?(payload, "speedAndDistance")
  end

  test "file name ends with json", %{exporter: exporter} do
    assert {:ok, content} = JSONExporter.export_async(exporter, Faker.valid_svd())
    assert String.ends_with?(content.file_name, ".json")
  end

  test "json round trip preserves key fields", %{exporter: exporter} do
    original = Faker.valid_svd()
    assert {:ok, content} = JSONExporter.export_async(exporter, original)
    assert {:ok, payload} = Jason.decode(content.data)

    assert get_in(payload, ["general", "imo"]) == original.general.imo
    assert get_in(payload, ["general", "shipName"]) == original.general.ship_name
    assert get_in(payload, ["general", "eventType"]) == original.general.event_type
  end

  test "json includes emissions section for dotnet/c++ parity", %{exporter: exporter} do
    assert {:ok, content} = JSONExporter.export_async(exporter, Faker.valid_svd())
    assert {:ok, payload} = Jason.decode(content.data)

    assert get_in(payload, ["emissions", "totalCo2"]) == Faker.valid_svd().emissions.total_co2
  end

  test "json exporter returns error when payload cannot be encoded", %{exporter: exporter} do
    svd = %StandardisedVesselDataset{
      general: Faker.valid_general_information(),
      cargo: %{__struct__: UnknownStruct, value: 1}
    }

    assert {:error, _} = JSONExporter.export_async(exporter, svd)
  end
end

File: src/go/svd/exporters/formats/json/json_exporter_test.go
package json_test

import (
	"context"
	"encoding/json"
	"strings"
	"testing"

	jsonexporter "github.com/WMS-Ceataec/standardised-vessel-dataset/src/go/svd/exporters/formats/json"
	"github.com/WMS-Ceataec/standardised-vessel-dataset/src/go/svd/internal/testdata"
	"github.com/WMS-Ceataec/standardised-vessel-dataset/src/go/svd/models"
	"github.com/WMS-Ceataec/standardised-vessel-dataset/src/go/svd/validators"
)

func TestJSONExporter(t *testing.T) {
	exporter := jsonexporter.NewJSONExporter(validators.NewStandardisedVesselDatasetValidator())

	t.Run("valid_export", func(t *testing.T) {
		result, err := exporter.ExportAsync(context.Background(), testdata.ValidSVD())
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if result.ContentType != "application/json" {
			t.Fatalf("unexpected content type: %s", result.ContentType)
		}
		if result.FileName != "SVD_9876543_2025-01-02.json" {
			t.Fatalf("unexpected file name: %s", result.FileName)
		}

		var decoded models.StandardisedVesselDataset
		if err := json.Unmarshal(result.Data, &decoded); err != nil {
			t.Fatalf("invalid json output: %v", err)
		}
		if decoded.General == nil || decoded.General.Imo != "9876543" {
			t.Fatalf("missing general imo in output: %+v", decoded.General)
		}
	})

	t.Run("invalid_svd", func(t *testing.T) {
		result, err := exporter.ExportAsync(context.Background(), testdata.InvalidSVD())
		if err == nil {
			t.Fatal("expected validation error")
		}
		if result != nil {
			t.Fatal("expected nil result")
		}
		if !strings.Contains(err.Error(), "Validation failed") {
			t.Fatalf("unexpected error: %v", err)
		}
	})

	t.Run("nil_svd", func(t *testing.T) {
		result, err := exporter.ExportAsync(context.Background(), nil)
		if err == nil || err.Error() != "svd cannot be nil" {
			t.Fatalf("unexpected error: %v", err)
		}
		if result != nil {
			t.Fatal("expected nil result")
		}
	})
}
